### PR TITLE
Migrate to Composer for dependency management and autoloading

### PR DESCRIPTION
**Pull request by Xenopathic**
* All files have been renamed to fit PSR-4, so the filename matches the class name including case.
* Composer is used for dependency management: see composer.json
* The autoloader has been removed and replaced with Composer's autoloader

Composer is awesome - its JSON file is easy to read, and getting the latest dependencies is slightly easier than with git. Additionally, 3rdparty files now go in /vendor rather than /3rdparty - apparently it's a convention that we should stick to.

TL;DR: Less code to maintain (removed autoloader), easier dependency management (Composer), more standard conformant.

### CODE REVIEW

**Comment by MightyCreak** [comment]
Nice! I wanted to see how Composer looked like.

I'll check this PR asap!

**Comment by MightyCreak** [comment]
I don't think it's necessary to add `composer.phar` into version control.
I installed Composer and git says that my `.phar` is different than the one in the code base. Also, one can install Composer globally.

**Comment by MightyCreak** [comment]
I have a problem running Mesamatrix, I get this error right when I enter the site:

```
<b>Fatal error</b>:  Uncaught exception 'InvalidArgumentException' with message 'A non-empty PSR-4 prefix must end with a namespace separator.' in /home/creak/code/web/mesamatrix/vendor/composer/ClassLoader.php:225
Stack trace:
#0 /home/creak/code/web/mesamatrix/vendor/composer/autoload_real.php(33): Composer\Autoload\ClassLoader-&gt;setPsr4('Mesamatrix', Array)
#1 /home/creak/code/web/mesamatrix/vendor/autoload.php(7): ComposerAutoloaderInitb13dfd2fc2a3956e078dd80dbc8f112b::getLoader()
#2 /home/creak/code/web/mesamatrix/lib/base.php(32): require('/home/creak/cod...')
#3 /home/creak/code/web/mesamatrix/lib/base.php(51): Mesamatrix::init()
#4 /home/creak/code/web/mesamatrix/http/index.php(21): require_once('/home/creak/cod...')
#5 {main}
  thrown in <b>/home/creak/code/web/mesamatrix/vendor/composer/ClassLoader.php</b> on line <b>225</b>
```

I'll try to see why is it happening, but I'm posting it here because maybe you already know the cause of the problem :]

**Comment by Xenopathic** [comment]
Ah, I thought that wasn't necessary. I know what the problem is :smile:

**Comment by Xenopathic** [comment]
As far as the phar is concerned: the official guidelines are to have a copy of it available in version control, so that those who don't want to install it can use it, and everything is packaged in the repository. There should be no reason why a freshly downloaded Composer should be different to that one though - I dl'ed and installed yesterday.

**Comment by MightyCreak** [comment]
I still have the same fatal error :/

As for the `composer.phar`, it makes sense ;)

**Comment by Xenopathic** [comment]
Ah no, that error you got was a broken json - I fixed it in the amended commit :smile:

**Comment by MightyCreak** [comment]
Ok it works, but apparently I'm still struggling with git :/  
What is the command line to get the latest code knowing that the history of the remote server has been modified?

I tried `git pull -f https://github.com/Xenopathic/mesamatrix.git composer` but I kept having merging issue... I ended up deleting my branch and pulling a new one, but I'm sure there's a better way...

**Comment by MightyCreak** [comment]
Last thing: apparently your last change in `composer.json` needs to update the lock file (I have a diff locally).

```diff
         "Read more about it at http://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file",
         "This file is @generated automatically"
     ],
-    "hash": "36a1096aab0c4c0d49dbd070bc386da6",
+    "hash": "7e113ead56fbe1bcad12a855d24c5e80",
     "packages": [
         {
             "name": "psr/log",
```

**Comment by Xenopathic** [comment]
@MightyCreak That is usually the easiest way to do a force pull, but another way (using more advanced commands, but you keep the branch) involves running `git fetch ...` (not `git pull ...`), followed by updating your branch by running `git reset --hard .../composer`. Replace `...` with the remote name.

**Comment by MightyCreak** [comment]
I updated the mesamatrix server.

Only problem... apparently UrlCache doesn't work anymore because I got timeout for each get_headers()... locally it works though. Do you think mesamatrix.net have been blacklisted by opengl.org because of the ~200 get_headers() that were done frequently a month ago?

**Comment by Xenopathic** [comment]
@MightyCreak That's a possible explanation for it :laughing: You sure your server isn't having some kind of connectivity issues?
